File: app/src/main/java/com/example/ama/questapp/presentation/list/di/ListModule.java
package com.example.ama.questapp.presentation.list.di;

import com.example.ama.questapp.presentation.list.QuestListFragment;
import com.example.ama.questapp.presentation.list.adapter.OnCompleteQuestClickListener;

import dagger.Module;
import dagger.Provides;

@Module
public class ListModule {
    private QuestListFragment fragment;

    public ListModule(QuestListFragment fragment) {
        this.fragment = fragment;
    }

    @Provides
    @MainQuestListScope
    OnCompleteQuestClickListener provideListener() {
        return fragment::tryToCompleteQuest;
    }

}

File: app/src/main/java/com/example/ama/questapp/di/modules/ListModule.kt
package com.example.ama.questapp.di.modules

import com.example.ama.questapp.presentation.list.QuestListFragment
import com.example.ama.questapp.presentation.list.adapter.OnCompleteQuestClickListener
import dagger.Module
import dagger.Provides

@Module
class ListModule {

    @Provides
    fun provideListener(fragment: QuestListFragment): OnCompleteQuestClickListener =
            OnCompleteQuestClickListener { fragment.tryToCompleteQuest(it) }

}